core/src/main/java/org/apache/calcite/rel/metadata/RelMdPredicates.java

```diff
@@ -562,34 +565,54 @@ public RelOptPredicateList getPredicates(Exchange exchange,
   /**
    * Infers predicates for a Values.
    *
-   * <p>The predicates on {@code T (x, y, z)} with rows
-   * {@code (1, 2, null), (1, 2, null), (5, 2, null)} are {@code 'y = 2'} and {@code 'z is null'}.
+   * <p>The predicates on {@code T (w, x, y, z)} with rows
+   * {@code (1, 2, 3, null), (1, 2, null, null), (5, 2, 3, null)} are
+   * {@code 'SEARCH($0, Sarg[1, 5])'},
+   * {@code '=($1, 2)'},
+   * {@code 'SEARCH($2, Sarg[3; NULL AS TRUE])'} and
+   * {@code '[IS NULL($3)'}.
    */
   public RelOptPredicateList getPredicates(Values values, RelMetadataQuery mq) {
     ImmutableList<ImmutableList<RexLiteral>> tuples = values.tuples;
     if (!tuples.isEmpty()) {
-      Set<Integer> constants = new HashSet<>();
-      IntStream.range(0, tuples.size()).boxed().forEach(constants::add);
-      List<RexLiteral> firstTuple = new ArrayList<>(tuples.get(0));
-      for (ImmutableList<RexLiteral> tuple : tuples) {
-        if (constants.isEmpty()) {
-          break;
-        }
-        for (int i = 0; i < tuple.size(); i++) {
-          if (!Objects.equals(tuple.get(i), firstTuple.get(i))) {
-            constants.remove(i);
-          }
+      List<RexLiteral> firstTuple = tuples.get(0);
+      List<TreeSet<RexLiteral>> valueList = new ArrayList<>();
+      for (int i = 0; i < firstTuple.size(); i++) {
+        valueList.add(
+            i, new TreeSet<>((o1, o2) -> o1.equals(o2) ? 0 : 1));
```

Comment:
If we use plain equals do we need to use a `TreeSet`? The code implies that we don't.

Reply:
Yes. I will update the code. The TreeSet have a getFirst method since 21. It is not necessary now. I will use HashSet.